lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,253 @@
+"""
+Models used for robust grading.
+
+Robust grading allows student scores to be saved per-subsection independent
+of any changes that may occur to the course after the score is achieved.
+"""
+
+from django.db import models
+from model_utils.models import TimeStampedModel
+
+from coursewarehistoryextended.fields import UnsignedBigIntAutoField
+from opaque_keys.edx.locator import BlockUsageLocator
+from xmodule_django.models import CourseKeyField, UsageKeyField
+
+from hashlib import sha256
+import json
+
+
+class VisibleBlocks(models.Model):
+    """
+    A django model used to track the state of a set of visible blocks under a given subsection at the time they are
+    used for grade calculation.
+
+    This state is represented using an array of serialized BlockRecords, stored in the blocks_json field. A
+    hash of this json array is used for lookup purposes.
+    """
+    _blocks_json = models.TextField(db_column="blocks_json")
+    hashed = models.CharField(max_length=32, primary_key=True)
+
+    def __unicode__(self):
+        """
+        String representation of this model.
+        """
+        return self.hashed
+
+    @classmethod
+    def create(cls, blocks):
+        """
+        Creates a new VisibleBlocks model. Argument 'blocks' should be an iterable collection of BlockRecords.
+        """
+        # Start by sorting the blocks to ensure equality regardless of how the blocks are ordered
+        sorted_blocks = sorted(
+            blocks,
+            key=lambda block: '{}{}{}'.format(block.block_key, block.max_score, block.weight)
+        )
+        _blocks_json = json.dumps(
+            [
+                block.to_dict()
+                for block in sorted_blocks
+            ]
+        )
+        hashed = sha256(_blocks_json).hexdigest()
```

Comment:
It's great that you're looking into the performance and security aspects of this. :-) As you folks have pointed out, the difference in CPU time between different hashing algorithms isn't very significant relative to the rest of the request. I suspect that the biggest difference you'll see between them is what happens at the database layer.

A normal int primary key is 4 bytes, and a bigint is 8 bytes. A SHA1 hash is 20 bytes. That's 40 bytes if we naively store the hexdigest, but we can also base64 encode that into a 28 byte varchar. Also, any secondary index (like a foreign key) is going to need to store it. This might add up for a particularly large table like `PersistentSubsectionGrade`, where we might see 100M+ entries.

I'm not an expert on MySQL by any means, so I'm mostly basing this on my interpretation of:
    http://dev.mysql.com/doc/refman/5.7/en/innodb-index-types.html

FWIW, I don't think this will make a huge difference for the scales that we're running at. But my inclination would be to do a simple primary key for this table and put a separate unique constraint on the hashed value, using the base64 encoded version. Doing it that way would also allow us to sidestep Django oddities around handling non-integer primary keys, like admin page functionality breaking in odd ways (this has bitten us before, though I forget the specifics).